x-pack/plugin/ml-package-loader/src/main/java/org/elasticsearch/xpack/ml/packageloader/action/ModelImporter.java

```diff
@@ -136,17 +261,58 @@ private void uploadVocabulary() throws URISyntaxException {
             true
         );
 
-        executeRequestIfNotCancelled(PutTrainedModelVocabularyAction.INSTANCE, request);
+        client.execute(PutTrainedModelVocabularyAction.INSTANCE, request, countingListener.acquire(r -> {
+            logger.debug(() -> format("[%s] imported model vocabulary [%s]", modelId, config.getVocabularyFile()));
+        }));
     }
 
-    private <Request extends ActionRequest, Response extends ActionResponse> void executeRequestIfNotCancelled(
-        ActionType<Response> action,
-        Request request
-    ) {
-        if (task.isCancelled()) {
-            throw new TaskCancelledException(format("task cancelled with reason [%s]", task.getReasonCancelled()));
+    private void indexPart(int partIndex, int totalParts, long totalSize, BytesArray bytes) {
+        PutTrainedModelDefinitionPartAction.Request modelPartRequest = new PutTrainedModelDefinitionPartAction.Request(
+            modelId,
+            bytes,
+            partIndex,
+            totalSize,
+            totalParts,
+            true
+        );
+
+        client.execute(PutTrainedModelDefinitionPartAction.INSTANCE, modelPartRequest).actionGet();
```

Comment:
This is the part that is blocking and/or slowing down the indexing, correct? We'll wait for the client response rather than continue asynchronously?

Reply:
The gain is from using 5 threads to stream the download and index the parts. The non-blocking write meant that we had more than 5 in flight requests (the download is faster than the indexing) and that was causing the OOM. In order to limit the number of requests to at most 5 there has to be some element of blocking. Model download uses it's a dedicated thread pool so the block does not starve other parts of the code of resources